### DISCUSSION TITLE
Transferring atom properties after a reaction

### DISCUSSION THREAD

**Discussion by cbouy** [rdkit/rdkit#4021] [Q&A]
Hi,

I was wondering if there was an efficient way to map back atom properties after a reaction, from the reactant to the product.

Code example:
```python
from rdkit import Chem
from rdkit.Chem import AllChem
mol = Chem.MolFromSmiles("CC(=O)[O-]")
mol.GetAtomWithIdx(3).SetProp("foo", "bar")
rxn = AllChem.ReactionFromSmarts("[C:1](=[O:2])-[O-:3]>>[C:1](-[O-:2])=[O+0:3]")
prod = rxn.RunReactants((mol,))[0][0]
[atom.GetPropsAsDict(True, False) for atom in prod.GetAtoms()]
# [{'old_mapno': 1, 'react_atom_idx': 1},
#  {'_QueryFormalCharge': -1, 'old_mapno': 2, 'react_atom_idx': 2},
#  {'_QueryFormalCharge': 0, 'old_mapno': 3, 'react_atom_idx': 3},
#  {'react_atom_idx': 0}]
```

The `foo` atom property is not copied to the product, so my current way of transferring it is the following:
```python
def reassign_props_after_reaction(reactant, product):
    """Maps back atomic properties from the reactant to the product.
    The product molecule is modified inplace.
    """
    for atom in product.GetAtoms():
        try:
            atom.GetIntProp("old_mapno")
        except KeyError:
            pass
        else:
            atom.ClearProp("old_mapno")
            idx = atom.GetUnsignedProp("react_atom_idx")
            old_atom = reactant.GetAtomWithIdx(idx)
            for prop, value in old_atom.GetPropsAsDict(True, False).items():
                set_atom_property(atom, prop, value)
        atom.ClearProp("react_atom_idx")

def set_atom_property(atom, attr, value):
    if isinstance(value, float):
        atom.SetDoubleProp(attr, value)
    elif isinstance(value, int):
        atom.SetIntProp(attr, value)
    elif isinstance(value, str):
        atom.SetProp(attr, value)

reassign_props_after_reaction(mol, prod)
[atom.GetPropsAsDict(True, False) for atom in prod.GetAtoms()]
# [{}, {'_QueryFormalCharge': -1}, {'_QueryFormalCharge': 0, 'foo': 'bar'}, {}]
```

Since in my case there are several reactions, several properties, and I use this on large molecules (it's part of the RDKitConverter that I developed for MDAnalysis), this part slows down the code quite significantly, so I'm looking for ways to improve that...

Thanks for any help!
Cedric

**Comment by bp-kelley** [comment]
Cedric,
   I expect most of the slowness is due to GetPropsAsDict, this is doing an
awful lot of work and is known to be slow.  I expect if you factored this
out it might get faster.

Alternatively, the C++ has a function atom.updateProps(other_atom) which I
expect would be considerably faster.  This, unfortunately, isn't exposed to
python but could be done easily.

Also atoms that aren't specified in the reaction, do have their properties
copied over and I would be ok copying all properties over to the product,
but i'll have to talk to Greg about this.

Cheers,
 Brian

**Reply by greglandrum** [reply]
@bp-kelley : having the reaction code copy over atom properties from template atoms makes sense to me. I can't think of anywhere that would cause problems, can you?

**Reply by mwojcikowski** [reply]
@greglandrum I agree with @cbouy that transferring props from **reactants** would be very useful feature. I consider reactions SMARTS to be a routing information for reactants, so should not really affect the contents much.

There is an obvious caveat, where SMARTS not only route but also introduce atoms. Although then there are no properties to be copied in the first place.

**Reply by cbouy** [reply]
Well that would be great! I'm guessing "computed" properties might be problematic (i.e. things like `_CIPRank`), unless they are overwritten afterwards?
And thanks for the tip with `GetPropsAsDict` @bp-kelley, I didn't know it was slow!

**Comment by cbouy** [comment]
Just tested the `RunReactantInPlace` from version 2021.09.3 on this, it works great!
Haven't tested it for performance on large molecules yet, but it will likely give it a huge boost!
Thanks a lot Greg 👍
